source/lac/sparsity_tools.cc

```diff
@@ -925,136 +925,67 @@ namespace SparsityTools
-    }
-
-    std::vector<MPI_Request> requests(send_data.size());
-
-
-    // send data
-    {
-      unsigned int idx = 0;
-      for (const auto &sparsity_line : send_data)
-        {
-          const int ierr =
-            MPI_Isend(DEAL_II_MPI_CONST_CAST(sparsity_line.second.data()),
-                      sparsity_line.second.size(),
-                      DEAL_II_DOF_INDEX_MPI_TYPE,
-                      sparsity_line.first,
-                      124,
-                      mpi_comm,
-                      &requests[idx++]);
-          AssertThrowMPI(ierr);
-        }
-    }
+    const auto receive_data = Utilities::MPI::some_to_some(mpi_comm, send_data);
 
-    {
-      // receive
-      std::vector<DynamicSparsityPattern::size_type> recv_buf;
-      for (unsigned int index = 0; index < num_receive; ++index)
-        {
-          MPI_Status status;
-          int        len;
-          int ierr = MPI_Probe(MPI_ANY_SOURCE, MPI_ANY_TAG, mpi_comm, &status);
-          AssertThrowMPI(ierr);
-          Assert(status.MPI_TAG == 124, ExcInternalError());
-
-          ierr = MPI_Get_count(&status, DEAL_II_DOF_INDEX_MPI_TYPE, &len);
-          AssertThrowMPI(ierr);
-          recv_buf.resize(len);
-          ierr = MPI_Recv(recv_buf.data(),
-                          len,
-                          DEAL_II_DOF_INDEX_MPI_TYPE,
-                          status.MPI_SOURCE,
-                          status.MPI_TAG,
-                          mpi_comm,
-                          &status);
-          AssertThrowMPI(ierr);
-
-          std::vector<DynamicSparsityPattern::size_type>::const_iterator ptr =
-            recv_buf.begin();
-          std::vector<DynamicSparsityPattern::size_type>::const_iterator end =
-            recv_buf.end();
-          while (ptr != end)
-            {
-              DynamicSparsityPattern::size_type num = *(ptr++);
-              Assert(ptr != end, ExcInternalError());
-              DynamicSparsityPattern::size_type row = *(ptr++);
-              for (unsigned int c = 0; c < num; ++c)
-                {
-                  Assert(ptr != end, ExcInternalError());
-                  dsp.add(row, *ptr);
-                  ++ptr;
-                }
-            }
-          Assert(ptr == end, ExcInternalError());
-        }
-    }
-
-    // complete all sends, so that we can safely destroy the buffers.
-    if (requests.size())
+    // add what we received
+    for (const auto &data : receive_data)
       {
-        const int ierr =
-          MPI_Waitall(requests.size(), requests.data(), MPI_STATUSES_IGNORE);
-        AssertThrowMPI(ierr);
+        const auto &recv_buf = data.second;
+        auto        ptr      = recv_buf.begin();
+        const auto  end      = recv_buf.end();
+        while (ptr != end)
+          {
+            const DynamicSparsityPattern::size_type num = *(ptr++);
+            Assert(ptr != end, ExcInternalError());
+            const DynamicSparsityPattern::size_type row = *(ptr++);
```

Comment:
while you're here (and in line 959/960), would you mind reverting the order of the first two elements? The `row` should precede the `n_entries` number because, conceptually, the latter is subsequent to the former (it's the number of entries *for this row*).